src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -173,6 +175,11 @@ private bool WaitForSignal(int timeoutMs)
                     }
 
                     counts = countsBeforeUpdate;
+                    if (timeoutMs != -1) {
+                        int waitMs = elapsedTicks - startTicks;
+                        if (waitMs <= timeoutMs)
+                            timeoutMs -= waitMs;
```

Comment:
Maybe something like this instead would avoid a remote chance of `waitMs` being negative:

```c#
uint waitMs = (uint)(endTicks - startTicks);
if (waitMs < (uint)timeoutMs)
    timeoutMs -= (int)waitMs;
else
    timeoutMs = 0;
```

Also could update `if (!WaitCore(timeoutMs))` at the top of the loop to `if (timeoutMs == 0 || !WaitCore(timeoutMs))` to avoid waiting again with a zero timeout. Maybe similarly for wasm.

Reply:
Could just check for negative `waitMs` too, may be simpler:

```c#
int waitMs = endTicks - startTicks;
if (waitMs >= 0 && waitMs < timeoutMs)
    timeoutMs -= waitMs;
else
    timeoutMs = 0;
```